moto/shield/models.py

```diff
@@ -0,0 +1,202 @@
+        self.tag_resource(protection.protection_arn, tags)
+        return protection.protection_id
+
+    def describe_protection(self, protection_id: str, resource_arn: str) -> Protection:  # type: ignore[return]
+        if protection_id and resource_arn:
+            msg = "Invalid parameter. You must provide one value, either protectionId or resourceArn, but not both."
+            raise InvalidParameterException(msg)
+
+        if resource_arn:
+            for protection in self.protections.values():
+                if protection.resource_arn == resource_arn:
+                    return protection
+            raise ResourceNotFoundException("The referenced protection does not exist.")
+
+        if protection_id:
+            if protection_id not in self.protections:
+                raise ResourceNotFoundException(
+                    "The referenced protection does not exist."
+                )
+            return self.protections[protection_id]
+
+    def list_protections(self, inclusion_filters: Dict[str, str]) -> List[Protection]:
+        """
+        Pagination has not yet been implemented
+        """
+        resource_protections = []
+        name_protections = []
+        type_protections = []
+
+        if inclusion_filters:
+            resource_arns = inclusion_filters.get("ResourceArns")
+            if resource_arns:
+                if len(resource_arns) > 1:
+                    raise ValidationException(
+                        "Error validating the following inputs: inclusionFilters.resourceArns"
+                    )
+                resource_protections = [
+                    protection
+                    for protection in self.protections.values()
+                    if protection.resource_arn == resource_arns[0]
+                ]
+            protection_names = inclusion_filters.get("ProtectionNames")
+            if protection_names:
+                if len(protection_names) > 1:
+                    raise ValidationException(
+                        "Error validating the following inputs: inclusionFilters.protectionNames"
+                    )
+                name_protections = [
+                    protection
+                    for protection in self.protections.values()
+                    if protection.name == protection_names[0]
+                ]
+            resource_types = inclusion_filters.get("ResourceTypes")
+            if resource_types:
+                if len(resource_types) > 1:
+                    raise ValidationException(
+                        "Error validating the following inputs: inclusionFilters.resourceTypes"
+                    )
+                type_protections = [
+                    protection
+                    for protection in self.protections.values()
+                    if protection.resource_type == resource_types[0]
+                ]
+            try:
+                protections = list(
```

Comment:
Can we change this to:

```
protections = []
protections.extend(type_protections)
#two more...
protections = list(set(protections))
```
I think that's logically equivalent? Less fancy, but more readable IMO.

Reply:
The above logic  fails this test : **test_list_protections_with_resource_arn_and_protection_name():**.  It gives me "AssertionError: assert 2 == 1". According to this: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/shield/client/list_protections.html
**Shield Advanced returns protections that exactly match all of the filter criteria that you provide.**. That means the resources listed should satisfy all the filter conditions in the **InclusionFilters**.

Reply:
Ah, gotcha, I misunderstood the logic